# 에러 & 예외 처리

:pushpin: [예외의 클래스 계층 구조](https://docs.python.org/ko/3/library/exceptions.html#exception-hierarchy)

## 에러(Error)

* 문법 에러(Syntax Error)

  * 문법 에러가 있는 프로그램은 실행이 안 된다.

  ```python
          ^  # 캐럿(caret)기호
  SyntaxError: invalid syntax
  ```

* 예외(Exception)
  * 실행 도중 Exception을 맞이하면, 실행을 멈춘다.
  * 문법적으로는 옳지만, 실행시 발생하는 에러
  * `ZeroDivisionError`, `NameError`, `TypeError`, `IndexError`, `KeyError` 등등

## 예외 처리(Exception Handling)

> `try` 문을 이용하여 예외 처리를 할 수 있다.

### Try & except

```python
try:
    <코드 블럭 1>
except (예외):
    <코드 블럭 2>
```

* 복수의 예외처리

```python
try:
    <코드 블럭 1>
except (예외1, 예외2):
    <코드 블럭 2>


try:
    <코드 블럭 1>
except 예외1:
    <코드 블럭 2>
except 예외2:
    <코드 블럭 3>
```

* **에러가 순차적으로 수행됨**으로, 가장 작은 범주부터 시작해야 한다.

```python
try:
    num = input('값을 입력하시오: ')
    100/int(num)
except Exception: # Exception 은 가장 큰 범주
    print('모르겠지만 에러야')
except ValueError:
    print('숫자를 넣어')
```

### else

> 에러가 발생하지 않는 경우 수행되는 문장은 `else`를 이용
>
> 모든 except 절 뒤에와야 한다.

```python
try:
    <코드 블럭 1>
except 예외:
    <코드 블럭 2>
else:
    <코드 블럭 3>
```

### finally

> 반드시 수행해야하는 문장에 활용
>
> 예외의 발생 여부과 관계없이 try 문을 떠날 때 항상 실행

```python
try:
    <코드 블럭 1>
except 예외:
    <코드 블럭 2>
finally:
    <코드 블럭 3>
```

### as

> `as` 키워드를 활용하여 에러 메시지를 보여줄 수 있다.

```python
try:
    <코드 블럭 1>
except 예외 as err:
    <코드 블럭 2>
```

### raise

> 예외를 강제로 발생시킨다. (Exception Raising)

```python
raise <에러>('메시지')
```

### assert

> 보통 **상태를 검증하는데 사용**되며 무조건 `AssertionError`가 발생한다.
>
> 일반적으로 디버깅용도로 사용